clients/src/main/java/org/apache/kafka/clients/producer/internals/StickyPartitioner.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.producer.internals;
+
+import org.apache.kafka.common.Cluster;
+import org.apache.kafka.common.PartitionInfo;
+
+import java.util.List;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.kafka.common.utils.Utils;
+
+public class StickyPartitioner {
+    private final ConcurrentMap<String, AtomicInteger> indexCache;
+    public StickyPartitioner() {
+        this.indexCache = new ConcurrentHashMap<>();
+    }
+
+    public int partition(String topic, Cluster cluster) {
+        AtomicInteger part = indexCache.get(topic);
+        if (part == null) {
+            return nextPartition(topic, cluster);
+        }
+        return part.get();
+    }
+
+    public int nextPartition(String topic, Cluster cluster) {
+        List<PartitionInfo> partitions = cluster.partitionsForTopic(topic);
+        int numPartitions = partitions.size();
+        List<PartitionInfo> availablePartitions = cluster.availablePartitionsForTopic(topic);
+        Integer random = Utils.toPositive(ThreadLocalRandom.current().nextInt());
+        AtomicInteger part = new AtomicInteger();
```

Comment:
I think there may have been a good reason why this used AtomicInteger in the past, but I am not sure there's any need for it now. I think you should just be able to use a `ConcurrentHashMap<String, Integer>`, given you're currently putting it back into the map after choosing it.